### PR TITLE
Expose the ClusterInfo object in the allocation explain output

### LINKED ISSUE #14405: Expose ClusterInfo in nodes' stats
Currently the only way to diagnose/debug issues with the gathered cluster info (disk and shard sizes) is to turn on verbose logging for the `cluster` namespace.

We should expose this in the nodes stats API so people can see the amount of free/used disk that is measured by Elasticsearch.

### PR DESCRIPTION
**Pull request by dakrone**
This adds an optional parameter to the cluster allocation explain API
that will return the cluster info object, `include_disk_info`, the
output looks like:

```
GET /_cluster/allocation/explain?include_disk_info -d'
{"index": "i", "shard": 0, "primary": false}'

{
  ... other info ...

  "cluster_info" : {
    "nodes" : {
      "7Uws-vL7R6WVm3ZwQA1n5A" : {
        "node_name" : "Kraven the Hunter",
        "least_available" : {
          "path" : "/path/to/data1",
          "total_bytes" : 165999570944,
          "used_bytes" : 118180614144,
          "free_bytes" : 47818956800,
          "free_disk_percent" : 28.8,
          "used_disk_percent" : 71.2
        },
        "most_available" : {
          "path" : "/path/to/data2",
          "total_bytes" : 165999570944,
          "used_bytes" : 118180614144,
          "free_bytes" : 47818956800,
          "free_disk_percent" : 28.8,
          "used_disk_percent" : 71.2
        }
      }
    },
    "shard_sizes" : {
      "[i][2][p]_bytes" : 0,
      "[i][4][p]_bytes" : 130,
      "[i][1][p]_bytes" : 0,
      "[i][3][p]_bytes" : 0,
      "[i][0][p]_bytes" : 130
    },
    "shard_paths" : {
      "[i][3], node[7Uws-vL7R6WVm3ZwQA1n5A], [P], s[STARTED], a[id=LegZLDniTVaw0Y1urv7s3g]" : "/path/to/data1/nodes/0",
      "[i][1], node[7Uws-vL7R6WVm3ZwQA1n5A], [P], s[STARTED], a[id=lAU_4vf_SKmoRdtg0ACnjQ]" : "/path/to/data1/nodes/0",
      "[i][2], node[7Uws-vL7R6WVm3ZwQA1n5A], [P], s[STARTED], a[id=Aurpeuj7SeGeyPDDpCtRgg]" : "/path/to/data1/nodes/0",
      "[i][0], node[7Uws-vL7R6WVm3ZwQA1n5A], [P], s[STARTED], a[id=Vgg8GlQTQ82C2j6HYBq8DQ]" : "/path/to/data1/nodes/0",
      "[i][4], node[7Uws-vL7R6WVm3ZwQA1n5A], [P], s[STARTED], a[id=t8hQlVSxQe-58fSeaXcAqg]" : "/path/to/data1/nodes/0"
    }
  }
}
```

Resolves #14405

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/cluster/DiskUsage.java [hunk 1]
```diff
@@ -44,6 +51,40 @@ public DiskUsage(String nodeId, String nodeName, String path, long totalBytes, l
...
+        out.writeVLong(this.freeBytes);
+    }
+
+    public XContentBuilder toShortXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field("path", this.path);
+        builder.byteSizeField("total_bytes", "total", this.totalBytes);
+        builder.byteSizeField("used_bytes", "used", this.getUsedBytes());
+        builder.byteSizeField("free_bytes", "free", this.freeBytes);
+        builder.field("free_disk_percent", this.getFreeDiskAsPercentage());
+        builder.field("used_disk_percent", this.getUsedDiskAsPercentage());
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/DiskUsage.java]: I wonder if these should just be rounded with `Math.round`? Do we really need fractional accuracy here, and then they can be indexed efficiently as integers?

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/DiskUsage.java]: rounding sounds fine to me, what do you suggest for the precision?

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/DiskUsage.java]: I'm suggesting to just round it to the nearest integer with `Math.round`.

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/DiskUsage.java]: I think that's too little precision, if you set your watermark at 90% and usage is at 89.9% you have no way of knowing if you are actually over or under the limit. How do you feel about a bit more precision?

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/DiskUsage.java]: No matter what precision you pick will have that problem, that is rounding `89.99` to one decimal point will round to `90`, and so on.

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/DiskUsage.java]: > No matter what precision you pick will have that problem, for example rounding 89.99 to one decimal point will round to 90, and so on.

Sure, but it's a bigger difference in terms of space, if we rounded to integer a 1% difference is ~10gb (assuming a 1tb disk), which is still quite a bit of space, if we rounded to one decimal it becomes ~1gb, or ~100mb for two decimals. 10gb per percentage is quite a coarse grain

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/DiskUsage.java]: I'm good with one decimal point with the caveat that this endpoint really should not be being indexed.

Comment (clintongormley) [inline core/src/main/java/org/elasticsearch/cluster/DiskUsage.java]: `half_float`?

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/DiskUsage.java]: I truncated this to a single decimal point

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java [hunk 2]
```diff
@@ -134,6 +135,15 @@ public boolean includeYesDecisions() {
         return this.includeYesDecisions;
     }
 
+    public void includeDiskInfo(boolean includeDiskInfo) {
+        this.includeDiskInfo = includeDiskInfo;
+    }
+
+    /** Returns true if information about disk usage and shard sizes should also be return */
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java]: Nit: `return` -> `returned`

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java [hunk 3]
```diff
@@ -134,6 +135,15 @@ public boolean includeYesDecisions() {
         return this.includeYesDecisions;
     }
 
+    public void includeDiskInfo(boolean includeDiskInfo) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java]: Should it have a Javadoc?

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequestBuilder.java [hunk 4]
```diff
@@ -53,6 +53,16 @@ public ClusterAllocationExplainRequestBuilder setPrimary(boolean primary) {
         return this;
     }
 
+    public ClusterAllocationExplainRequestBuilder setIncludeYesDecisions(boolean includeYesDecisions) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequestBuilder.java]: Javadocs?

File: core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequestBuilder.java [hunk 5]
```diff
@@ -53,6 +53,16 @@ public ClusterAllocationExplainRequestBuilder setPrimary(boolean primary) {
...
+        request.includeYesDecisions(includeYesDecisions);
+        return this;
+    }
+
+    public ClusterAllocationExplainRequestBuilder setIncludeDiskInfo(boolean includeDiskInfo) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequestBuilder.java]: Javadocs?

**Comment by jasontedor** [comment]
Modulo some nits, LGTM. Fire at will.

**Pull request merged**